services/galley/src/Galley/API/Message.hs

```diff
@@ -144,6 +147,22 @@ checkMessageClients sender participantMap recipientMap mismatchStrat =
         mkQualifiedMismatch reportedMissing redundant deleted
       )
 
+getRemoteClients :: ConvId -> Galley (Map (Domain, UserId) (Set ClientId))
+getRemoteClients convId = do
+  remoteMembers <- Data.lookupRemoteMembers convId
+  -- FUTUREWORK: parallelise RPCs
+  fmap mconcat -- concatenating maps is correct here, because their sets of keys are disjoint
+    . traverse (uncurry getRemoteClientsFromDomain)
```

Comment:
You could skip the `unTagged`, `Map.Assocs` call here by using `partitionRemote` instead of `partitionQualified`. And while you're at it, parallelize the requests as done here: https://github.com/wireapp/wire-server/blob/develop/services/galley/src/Galley/API/Query.hs#L106-L111